### PR TITLE
[jvm-packages] XGBoost Spark supports ranking with group data.

### PR DESCRIPTION
**Pull request by yanboliang**
XGBoost Spark supports ranking with group data.

### CODE REVIEW

**Comment by CodingCat** [comment]
eh....large than I expect...will look at it tmr afternoon (Friday afternoon :-) )

**Comment by hcho3** [comment]
@CodingCat @yanboliang Sorry for hijacking this thread, but would #2748 be useful for ranking tasks on Spark?

**Comment by yanboliang** [comment]
@hcho3 Yep, it would be useful for ranking on xgboost-spark, we are in the same direction. This PR expose a new group data API for xgboost-spark, we can update internal implementation if the backend C++ code changed. Thanks.

**Comment by hcho3** [comment]
@yanboliang That's good to know. I will add some tests to #2749 and merge it. Thanks!

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala [hunk 1]
```diff
@@ -21,12 +21,10 @@ import java.nio.file.Files
 
 import scala.collection.mutable
 import scala.util.Random
-
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: can we keep these empty lines to separate xgboost4j and the other imports?

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala [hunk 2]
```diff
@@ -56,8 +54,8 @@ object XGBoost extends Serializable {
   private val logger = LogFactory.getLog("XGBoostSpark")
 
   private def removeMissingValues(
-      denseLabeledPoints: Iterator[XGBLabeledPoint],
-      missing: Float): Iterator[XGBLabeledPoint] = {
+      denseLabeledPoints: Seq[XGBLabeledPoint],
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: if you take sequence, that means you will load a partition into memory entirely which will leads to OOM potentially

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala [hunk 3]
```diff
@@ -129,7 +127,7 @@ object XGBoost extends Serializable {
       rabitEnv.put("DMLC_TASK_ID", taskId)
       Rabit.init(rabitEnv)
       val watches = Watches(params,
-        removeMissingValues(labeledPoints, missing),
+        removeMissingValues(labeledPoints.toSeq, missing),
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: yeah, this toSeq is risky to lead to OOM

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala [hunk 4]
```diff
@@ -308,9 +306,26 @@ private class Watches private(
 
 private object Watches {
 
+  def formatGroups(groups: Seq[Int]): Seq[Int] = {
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: buildGroups or formGroups as a better name?

**Comment by CodingCat** [comment]
LGTM

**Comment by CodingCat** [comment]
can you file a PR from spark_dev_do_not_delete to master to ensure that they count as your contribution

**Pull request merged**